src/puya/ir/builder/assignment.py

```diff
@@ -75,26 +98,43 @@ def _handle_assignment(
                     " which is being passed by reference",
                     assignment_location,
                 )
+            if isinstance(var_type, wtypes.WTuple):
+                var_names = _build_tuple_names(var_name, var_type, var_loc)
+            else:
+                var_names = [(var_name, var_loc)]
             return assign(
                 context,
                 source=value,
-                names=[(var_name, var_loc)],
+                names=var_names,
                 source_location=assignment_location,
             )
         case awst_nodes.TupleExpression() as tup_expr:
             source = context.visitor.materialise_value_provider(
                 value, description="tuple_assignment"
             )
             items = lvalue_items(tup_expr)
-            if len(source) != len(items):
+            if len(source) != sum(_get_wtype_arity(i.wtype) for i in items):
                 raise CodeError("unpacking vs result length mismatch", assignment_location)
-            return [
-                val
-                for dst, src in zip(items, source, strict=True)
-                for val in handle_assignment(
-                    context, target=dst, value=src, assignment_location=assignment_location
+
+            results = list[Value]()
+            offset = 0
+            for item in items:
```

Comment:
I'm sure there's a nicer way to do this, but the goal is to match the arity of a assignment target to a slice of the source values (eg. tuple of len == 3 will consume 3 values)